pyat/at/tracking/particles.py

```diff
@@ -84,8 +56,80 @@ def _compute_bunch_length_from_espread(ring, espread):
     return blength
 
 
-def sigma_matrix(ring=None, twiss_in=None, emitx=None,
-                 emity=None, blength=None, espread=None):
+def _sigma_matrix_from_R66(R66, emitx, emity, blength, espread):
+    sig_matrix = emitx*R66[0] + emity*R66[1] + \
+             blength*espread*R66[2]
+    return sig_matrix
+
+
+def _sigma_matrix_lattice(ring=None, twiss_in=None, emitx=None,
+                          emity=None, blength=None, espread=None,
+                          verbose=False):
+
+    flag = emitx or emity or blength or espread
+    if flag:
+        assert emitx is not None, 'emitx must be defined'
+        assert emity is not None, 'emity must be defined'
+        assert blength is not None, 'blength must be defined'
+        assert espread is not None, 'espread must be defined'
+
+    if ring:
+        cavPassFlag = numpy.any(
+                        numpy.array(
+                          [i.PassMethod == 'CavityPass' for i in ring]))
+        radPassFlag = numpy.any(
+                        numpy.array(
+                          ['Rad' in i.PassMethod for i in ring]))
+
+        if cavPassFlag and not radPassFlag and not flag:
+            raise AtError('Cannot compute 6D sigma matrix without '
+                          'radiation damping and without emittances. '
+                          'Either switch on radiation or provide the '
+                          'emittances.')
+
+        ld0, bd, ld = ring.get_optics()
+        if ring.radiation and not flag:
+            if verbose:
+                print('Generating correlated sigma matrix using '
+                      'ohmi envelope')
+
+            emit0, beamdata, emit = ohmi_envelope(ring, refpts=[0])
+            sig_matrix = emit.r66[0]
+        elif flag:
+            if verbose:
+                print('Generating pseudo-correlated matrix '
+                      'from initial optics conditions')
+            if ring.radiation:
```

Comment:
Why ignoring the bunch length if `ring.radiation==True`?
Alternatively, you may run this calculation if espread is provided but not blength

Reply:
If radiation is on, then there is a correlation between bunch length and energy spread. So in this case, the bunch length is computed based on the provided energy spread.